R/check_frontmatter.R

```diff
@@ -8,103 +8,166 @@
+#' session.
 #'
-#' @return If one of the checks fails, an error message will be returned.
-#' Otherwise, the parsed yml front matter if \code{yaml} is
-#' \code{TRUE} is returned, or a message that everything is OK otherwise.
+#' @return A report of all failed checks.
 #'
 #' @importFrom rmarkdown yaml_front_matter
-#' @importFrom assertthat assert_that is.string has_name
+#' @importFrom assertthat assert_that is.string has_name is.flag noNA
 #' @importFrom stringr str_detect
-#' @importFrom purrr map_lgl
+#' @importFrom purrr map_lgl map_chr
 #'
 #' @export
 #'
 #'
 check_frontmatter <- function(
-  protocol_code,
-  yaml = TRUE
-  ) {
-
-  path_to_protocol <- get_path_to_protocol(
-    protocol_code = protocol_code)
+    protocol_code,
+    fail = !interactive()
+) {
+  assert_that(str_detect(protocol_code, "^s[fioap]p-\\d{3}-(nl|en)$"))
+  assert_that(is.flag(fail), noNA(fail))
 
-  yml <- yaml_front_matter(file.path(path_to_protocol, "index.Rmd"))
+  x <- load_protocolcheck(x = protocol_code)
+  template_name <-
+    gsub(
+      pattern = "(s\\w{1}p)-\\d*-(\\w{2})",
+      replacement = "template_\\1_\\2",
+      protocol_code
+    )
+  path_to_template <-
+    system.file(
+      file.path("rmarkdown", "templates", template_name, "skeleton"),
+      package = "protocolhelper")
 
-  assert_that(has_name(yml, "title"))
-  assert_that(is.string(yml$title))
+  yml_protocol <- yaml_front_matter(input = file.path(x$path, "index.Rmd"))
+  yml_template <- yaml_front_matter(input = file.path(path_to_template,
+                                                      "skeleton.Rmd"))
 
-  if (has_name(yml, "subtitle")) {
-    assert_that(is.string(yml$subtitle))
-  }
+  # check if all yaml keys are present
+  yml_missing <- yml_template[!names(yml_template) %in% names(yml_protocol)]
+  yml_missing$subtitle <- NULL
+  problems <- sprintf(
+    "The yaml-key '%s' is missing",
+    names(yml_missing)
+  )
 
-  assert_that(has_name(yml, "author"))
-  assert_that(
-    all(
-      purrr::map_lgl(yml$author, ~has_name(., "name"))
-      ),
-    all(
-      purrr::map_lgl(yml$author, ~has_name(., "orcid"))
-    )
+  # checks common to all protocol types
+  yml_string <- list("title" = yml_protocol$title,
+                  "file_manager" = yml_protocol$file_manager)
+  problems <- c(problems,
+                sprintf(
+                  "'%s' must be a string",
+                  names(yml_string)[!map_lgl(yml_string, is.string)])
   )
-  assert_that(
-    all(
-      purrr::map_lgl(yml$author, ~is.string(.$name))
-    ),
-    all(
-      purrr::map_lgl(yml$author, ~is.string(.$orcid))
+
+  if (has_name(yml_protocol, "subtitle")) {
+    problems <- c(problems,
+                    paste0(
+                      "subtitle is not a string, NULL, or an empty string, ",
+                      "please remove in the yaml header if not needed."
+                    )[!is.string(yml_protocol$subtitle) |
+                        yml_protocol$subtitle == "" |
```

Comment:
I tested it again with a subtitle without text (or with an empty text) in the yaml header, but this still gave no error.  I think you have to remove the command below:
- in case of an empty subtitle entry, this gives not `TRUE` or `FALSE` but `logical(0)`, after which the problem is not added to the `protocolcheck` object
- I suppose the situation of an empty string for subtitle doesn't exist, at least I am not able to create this situation in a yaml heading...
```suggestion
```

Reply:
Good catch. See [b7ac618](https://github.com/inbo/protocolhelper/pull/71/commits/b7ac618ef35ea781815907f41e7456a51822c111).  The problem was not line 71. It's possible to create a subtitle with empty string (see updated unit test). I noticed `!is.string()` part will return `TRUE `also in the `NULL` case. So replacing `|` with `||` and removing the redundant `!is.null()` solved the problem.

Reply:
I might be missing something, but the unit test shows that a user can do `create_protocol(..., subtitle = "", ...)` (for now... I should probably catch that case in that function as well)?